### ISSUE TITLE
Message: (Operation not permitted) trying reading config file.

### ISSUE THREAD

**Issue by epsilon77** [adrienverge/openfortivpn#976]
Hello!

I am getting this error when I run openfortivpn 1.17.2 with sudo.

WARN:   Could not load configuration file "/etc/openfortivpn/config" (Operation not permitted).

![DeepinScreenshot_Seleccionar área_20220417120404](https://user-images.githubusercontent.com/991938/163725057-036160c2-8d17-4441-b153-1a9726ae708d.png)

however, I check the config file permissions and these seems fine. (see image), I tried changing permissions to rwx for all, but the same result.

Its normal this behaivor? (in the previous versions I don't see that error message)

Regards

**Comment by DimitriPapadopoulos** [comment]
This is not expected. This file might be read-protected by other mechanisms than mere POSIX access rights. You'll have to find what's going wrong.

Does the following work?
```
sudo cat /etc/openfortivpn/config
```
What is the output of the following?
```
ls -ltrh /etc/openfortivpn
df /etc/openfortivpn/config
```

**Comment by epsilon77** [comment]
Hello!

Sorry for late reply. 

![DeepinScreenshot_Seleccionar área_20220420204719](https://user-images.githubusercontent.com/991938/164355371-e83dfc64-1e34-4f4f-8e32-09412890f45d.png)

**Comment by DimitriPapadopoulos** [comment]
Strange indeed. If `sudo cat /etc/openfortivpn/config` works, I don't see why `sudo openfortivpn` wouldn't be able to read the file.

Function `load_config()` fails:
https://github.com/adrienverge/openfortivpn/blob/d2199f2eebc22b850225cb55c51dbb9e7a9799bb/src/main.c#L562-L567

Try `openfortivpn -v -v` to get extra debug messages.

**Comment by DimitriPapadopoulos** [comment]
But I doubt `openfortivpn -v -v` will help. You need to add extra print statements in the source code to find what fails, perhaps not `fopen()` itself:
https://github.com/adrienverge/openfortivpn/blob/d2199f2eebc22b850225cb55c51dbb9e7a9799bb/src/config.c#L190-L192
but `fstat()`:
https://github.com/adrienverge/openfortivpn/blob/d2199f2eebc22b850225cb55c51dbb9e7a9799bb/src/config.c#L194-L196
or `getline()`:
https://github.com/adrienverge/openfortivpn/blob/d2199f2eebc22b850225cb55c51dbb9e7a9799bb/src/config.c#L200

**Comment by DimitriPapadopoulos** [comment]
Also is this specific to version 1.17.2?

**Comment by DimitriPapadopoulos** [comment]
What's the output of the following command?
```
sudo stat /etc/openfortivpn/config
```

**Comment by epsilon77** [comment]
Hello!

sorry for the late reply.. yes I see the warning message only in 1.17.2 version,  when I tried 1.17.1 I don't get this message. Btw I tried in two different Debian machine. and I see the warning message in both, when I run 1.17.2 version.

output of sudo stat /etc/openfortivpn/config

![DeepinScreenshot_Seleccionar área_20220422221120](https://user-images.githubusercontent.com/991938/164872963-22edc2d5-05b5-41e2-8956-f25988f911a0.png)

**Comment by DimitriPapadopoulos** [comment]
OK, it's probably not `fopen()` or `fstat()` failing, since that part of the code hasn't changed and `cat` and `stat` work well too.

The root of the problem must be in the `fread()` → `getline()` transition in #952. Unfortunately, I cannot reproduce this problem on Ubuntu, and I cannot find anything suspect in the new code.

Note that the openfortivpn package on Ubuntu 20.04 is slightly different, with different access rights for `/etc/openfortivpn/config`:
```
$ ls -al /etc/openfortivpn/
total 20
drwxr-xr-x   2 root root  4096 Apr 27  2020 .
drwxr-xr-x 161 root root 12288 Apr 22 20:10 ..
-rw-r--r--   1 root root   151 Feb 27  2020 config
$ 
$ dpkg -S /etc/openfortivpn/config
openfortivpn: /etc/openfortivpn/config
$ 
$ dpkg -l openfortivpn
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version      Architecture Description
+++-==============-============-============-===============================================
ii  openfortivpn   1.12.0-1     amd64        Fortinet client for PPP+SSL VPN tunnel services
$ 
```

**Comment by DimitriPapadopoulos** [comment]
Does it work any better if you use a config file in a different location, for example outside `/etc`?

It would help if you could build the latest sources and debug [`load_config()`](https://github.com/adrienverge/openfortivpn/blob/49ad913622c415712049de36b06bfa3ef8f2eac4/src/config.c#L179), especially the last few lines:
https://github.com/adrienverge/openfortivpn/blob/49ad913622c415712049de36b06bfa3ef8f2eac4/src/config.c#L449-L465

Either add some print() statements or run openfortivpn in gdb and add a break point in this function.

**Comment by DimitriPapadopoulos** [comment]
It works for me using `/etc/openfortivpn/config` as a configuration file, with seemingly identical file properties:
```
# ls -al /etc/openfortivpn
total 20
drwx------   2 root root  4096 avril 23 09:34 .
drwxr-xr-x 161 root root 12288 avril 23 09:32 ..
-rw-------   1 root root   306 avril 23 09:34 config
# 
# 
# cat /etc/openfortivpn/config 
# ### config file for openfortivpn, see man openfortivpn(1) ###
# 
# host = vpn.example.org
# port = 443
# username = vpnuser
# password = VPNpassw0rd
# config file for openfortivpn, see man openfortivpn(1)
host = XXXXXXXXXXXXX.XXX.XX
port = 443
username = XXXXXXXX
half-internet-routes = 1
use-syslog = 0
# 
```
Which exact version of Debian is this?

**Comment by DimitriPapadopoulos** [comment]
Oh, it's really an issue with the `getline()` error detection in #952:

https://github.com/adrienverge/openfortivpn/blob/b070c0b7fcc3bc0e1f33988947637f7363e9f599/src/config.c#L457-L458

Indeed `getline()` will set `errno` in case of error, but it will **not reset** it in the absence of error. So we really need to reset `errno` ourselves before calling `getline()`. This is not clearly explained by the man page:
> Both functions return -1 on failure to read a line (including end-of-file condition).  In the event of an error, _errno_ is set to indicate the cause.

**Comment by DimitriPapadopoulos** [comment]
Actually, it might be better to rely on `feof()`/`ferror()` rather than `errno` to detect `getline()` errors. The [POSIX documentation](https://pubs.opengroup.org/onlinepubs/9699919799/functions/getline.html) is much clearer than the man page:
> Upon successful completion, the _getline_() and _getdelim_() functions shall return the number of bytes written into the buffer, including the delimiter character if one was encountered before EOF, but excluding the terminating NUL character. If the end-of-file indicator for the stream is set, or if no characters were read and the stream is at end-of-file, the end-of-file indicator for the stream shall be set and the function shall return -1. If an error occurs, the error indicator for the stream shall be set, and the function shall return -1 and set _errno_ to indicate the error.

**Comment by Totorokrut** [comment]
Hello.

Is it possible that this issue is back? :)

<img width="804" height="732" alt="Image" src="https://github.com/user-attachments/assets/c02f7b25-62e6-4c87-86af-43d8e972d65f" />

**Comment by Totorokrut** [comment]
Yes, `sudo cat` and `sudo head` works fine.

```
lglassner@G10:~$ sudo head -n2 /home/lglassner/config
### configuration file for openfortivpn, see man openfortivpn(1) ###

lglassner@G10:~$ sudo file /home/lglassner/config
/home/lglassner/config: ASCII text
 ```

**Comment by bodasadallah** [comment]
I have the same problem. The file exists with normal permissions, but it still can't be read.

EDITL: This only happens with the snap package, installing it using apt works fine

**Comment by DimitriPapadopoulos** [comment]
@bodasadallah This issue has been closed, it has been fixed by #979.

Which version of openfortivpn are you running? Have you tried 1.24 which includes #979?
